Allow crontab resource to read crontab at user specified paths.

miah opened the pull request:
This PR adds support for reading crontab files at user specified paths as requested in https://github.com/chef/inspec/issues/1461.

Example usage:
```
describe crontab('/etc/cron.d/some_crontab') do
  its('commands') { should include '/path/to/some/script' }
end
```

miah committed:
add a emulated /etc/cron.d/crondotd file to the mocking system.

miah committed:
test that we handle incoming paths correctly by rendering to_s.

miah committed:
We take in both users and a path, so lets call that destination.

miah committed:
To make the test pass we'll determine if we are dealing with a path or
a user and return the correct string.

miah committed:
we will need the ability to determine if we are dealing with a path when either calling the crontab command or reading the file directly, so break that out into a path? method.

miah committed:
remove author field.

miah committed:
test contents of our crondotd file.

miah committed:
we have to explicitly make @destination a String to use include?.

miah committed:
when we get a path we use inspec.file to get conents, otherwise we run the crontab command.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add documentation for example usage with file path.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -96,7 +100,16 @@ def crontab_cmd
+      @destination.nil? ? 'crontab for current user' : path_or_user
+    end
+
+    def path_or_user
```
We should probably make this a private method.

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -73,7 +77,7 @@ def parse_crontab_line(l)
 
     def crontab_cmd
-      @user.nil? ? 'crontab -l' : "crontab -l -u #{@user}"
+      @destination.nil? ? 'crontab -l' : "crontab -l -u #{@destination}"
```
I don't think this works... `-u` specifically takes a user. On a stock Ubuntu 16.04 box, which has a `/etc/cron.d/mdadm` file, `crontab -l -u /etc/cron.d/mdadm` fails. I think you just need to read in the file.

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -25,21 +24,26 @@ class Crontab < Inspec.resource(1)
     def read_crontab
-      inspec.command(crontab_cmd).stdout.lines.map { |l| parse_crontab_line(l) }.compact
+      ct = path? ? inspec.file(@destination).content : inspec.command(crontab_cmd).stdout
+      ct.lines.map { |l| parse_crontab_line(l) }.compact
```
Non-user crontabs, such as those in `/etc/cron.d` follow a different format. Specifically, before the command, there is another field for the user via which the command should be run. We'd need to adapt the resource to handle that new format for non-user crontabs.

adamleff commented on test/unit/mock/files/crondotd:
```diff
@@ -0,0 +1,10 @@
+#
```
I think we need to provide a better mock file that has users in it. Here's the contents of the `/etc/cron.d/mdadm` file from the Ubuntu vagrant box I just spun up:

```
#
# cron.d/mdadm -- schedules periodic redundancy checks of MD devices
#
# Copyright © martin f. krafft <madduck@madduck.net>
# distributed under the terms of the Artistic Licence 2.0
#

# By default, run at 00:57 on every Sunday, but do nothing unless the day of
# the month is less than or equal to 7. Thus, only run on the first Sunday of
# each month. crontab(5) sucks, unfortunately, in this regard; therefore this
# hack (see #380425).
57 0 * * 0 root if [ -x /usr/share/mdadm/checkarray ] && [ $(date +\%d) -le 7 ]; then /usr/share/mdadm/checkarray --cron --all --idle --quiet; fi
```

miah replied:
Thanks, my precise64 doesn't have this; even after I installed mdadm. =)

adamleff reviewed (changes requested):
Hi @miah! I love that we're enhancing this resource. However, I think some incorrect assumptions were made regarding the format of non-user crontab files and how they can be read in. Perhaps I tested on a different OS than you did, so I'm happy to be wrong about this as well.

chris-rock commented on lib/resources/crontab.rb:
```diff
@@ -25,21 +24,26 @@ class Crontab < Inspec.resource(1)
 
-    def initialize(user = nil)
-      @user   = user
+    def initialize(destination = nil)
```
I am a little bit unsure about the overloading of the destination. It can be a user name or a path. Wouldn't it be cleaner to use something like:
`describe crontab({path: '/etc/cron.d/some_crontab'})` and `describe crontab({user: 'chris'})`

miah replied:
Yes will do!

miah committed:
Make path? and path_or_user private methods

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add missing username filed to crondotd mock file

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Pass argument as a hash when testing file paths

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Expected results should include usernames when testing file paths

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add special string `@yearly` test to crondotd mock file

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add user to existing cron tests

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Rubocop says I need spaces after/before curly brackets

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add user to crondotd file tests and add @yearly test

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Modify initialize to take options hash and be backwards compatible.

Change initialize default argument to create a hash by default, though
it is still possible to pass in a 'user' string argument.

@user gets set with the argument value unless its a hash, in which case
it tries to set the value of the user key, otherwise it becomes nil.

@file gets set with the value of the path key, unless it doesn't exist
in which case it becomes nil.

All hash keys are symbolized to ensure consistent access.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Check if @path is nil to determine if we run crontab command or parse
file.

path? was removed as we're not overloading a @destination variable
anymore.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
if @user is nil assume current user otherwise crontab for @user

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Change to complete if rather than ternary.

We have three possible cases, current user, other user, or file path.
This accounts for all of them.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add user to the crontab FilterTable

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Remove path? and path_or_user

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Move crontab parsing to two methods, parse_user_crontab and
parse_system_crontab

Because a command in a crontab file could have spaces we must parse user
and system crontabs differently.

When we parse user crontabs the user field will either be nil, or the requested user.

Both user and path parsers handle special strings (@yearly, @weekly,
etc). And also account for position of user in these files (or adds it
in user case)

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Update examples with user: and path:

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add spaces after : in example docs

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Disable rubocop ClassLength check

Signed-off-by: Miah Johnson <miah@chia-pet.org>

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -1,15 +1,15 @@
 require 'utils/parser'
 require 'utils/filter'
 
+# rubocop:disable Metrics/ClassLength
```
To be consistent with the way we disable in other files, can you please move this to the class definition itself?

```ruby
class Crontab < Inspec.resource(1) # rubocop:disable Metrics/ClassLength
```

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -1,15 +1,15 @@
     desc 'Use the crontab InSpec audit resource to test the contents of the crontab for a given user which contains information about scheduled tasks owned by that user.'
     example "
-      describe crontab('root') do
+      describe crontab({user: 'root'}) do
```
The `{ }` braces aren't necessary, Ruby will happily see that you're passing in a hash if you use the 2.0-and-greater syntax like you're doing:

```ruby
describe crontab(user: 'root') do
```

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -25,51 +25,35 @@ class Crontab < Inspec.resource(1)
         its('entries.length') { should cmp 1 }
       end
+
+      describe crontab({path: '/etc/cron.d/some_crontab'}) do
```
Same here, unnecessary `{ }` characters :)

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -25,51 +25,35 @@ class Crontab < Inspec.resource(1)
-    def initialize(user = nil)
-      @user   = user
+    def initialize(opts = {})
+      Hash[opts.map { |k, v| [k.to_sym, v] }] if opts.respond_to?(:fetch)
```
We're doing a check to see if it's a Hash on each of these 3 lines, so can we consolidate it so it reads a bit more clearly?

```ruby
if opts.respond_to?(:fetch)
  # user supplied a Hash containing options
  @user = opts[:user]
  @path = opts[:path]
else
  # user passed in a string, and the prior behavior only supported
  # user crontabs, so assume they're supplying a user.
  @user = opts
  @path = nil
end
```

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -25,51 +25,35 @@ class Crontab < Inspec.resource(1)
 
     def read_crontab
-      inspec.command(crontab_cmd).stdout.lines.map { |l| parse_crontab_line(l) }.compact
+      ct = @path.nil? ? inspec.command(crontab_cmd).stdout : inspec.file(@path).content
```
We do the `@path.nil?` check a few times. What do you think about adding a helper method or to?

```ruby
def is_system_crontab?
  !@path.nil?
end

def is_user_crontab?
  !@user.nil?
end
```

We could use that here, in the `to_s` method, and in `parse_crontab_line`.

adamleff commented on lib/resources/crontab.rb:
```diff
@@ -25,51 +25,35 @@ class Crontab < Inspec.resource(1)
+      @path = opts.fetch(:path, nil) if opts.respond_to?(:fetch)
       @params = read_crontab
 
       return skip_resource 'The `crontab` resource is not supported on your OS.' unless inspec.os.unix?
```
While you're in here, can you change this to raise an exception?

```ruby
raise Inspec::Exceptions::ResourceSkipped, 'The `crontab` resource is not supported on your OS.' unless inspec.os.unix?
```

And we should also add an exception if the user didn't supply us enough information (i.e. they gave us a hash that contains neither user or path):

```ruby
raise Inspec::Exceptions::ResourceFailed, "A user or path must be supplied" if @user.nil? && @path.nil?
```

adamleff reviewed (changes requested):
@miah nice work! A couple of clean-up items and I think this will be ready to go.

miah committed:
Moved rubocop ClassLength metric next to class instead of above the
module.

Remove unnecessary braces.

Add is_system_crontab? and is_user_crontab helper methods and use them.

Add tests to see if error conditions are raised when the resource is
invoked with missing parameters (user, or path), and on a unsupported
os.

Change initialize to group all hash functions together and raise errors
when user and path is unset. Also raise errors on unsupported operating
systems.

Change order of ternary and use is_system_crontab? rather than
@path.nil?

Signed-off-by: Miah Johnson <miah@chia-pet.org>

adamleff reviewed (approved):
Alright, @miah! This looks great. Conditionally approved pending green tests after a rebase once we square away the unrelated Train issue.

arlimus commented on lib/resources/crontab.rb:
```diff
@@ -25,51 +24,40 @@ class Crontab < Inspec.resource(1)
         its('entries.length') { should cmp 1 }
       end
+
+      describe crontab(path: '/etc/cron.d/some_crontab') do
```
love the new options specification!

arlimus reviewed (approved):
What a wonderful improvement and update, thank you so much @miah !! 😄 

Kudos Adam for your guidance as well :)
